README.md
# crazy_list

Crazy list is a widget that change his layout depending of the item count on the list, can be square, triangle, diagonal, grid, list

## Usage

```dart
CrazyList(
  itemCount: itemCount,
  mode: CrazyListMode.grid,
  color1: Colors.green[50],
  color2: Colors.green[300],
  onTap: (index) => print('$index tapped'),
  builder: (context, index, mode) {
    return Padding(
      padding: const EdgeInsets.all(16.0),
      child: Text('My Label $index'),
    );
  },
),
```

By default depending of `itemCount` you'll get:
 
- Diagonal layout for item count 2
- Triangle layout for item count 3
- Square layout for item count 4
- List or Grid layout for item count more than 4

You can customize this widget with the following params: 

`mode` list or grid for item > 4

`color1` color 1 

`color2` color 2

`onTap` callback when an item is tapped
 
`builder` to build the item 

`itemCount` number of items

`curve` used for the enter animation

 

CHANGELOG.md
## [1.0.2] - 31/03/2019.

- animate list items on enter.

## [1.0.1] - 24/03/2019.

- center content on list / grid views.

## [1.0.0] - 23/03/2019.

- initial release.


lib/src/grid.dart
part of '../crazy_list.dart';

class GridWidget extends StatelessWidget {
  final WidgetBuilder builder;
  final int itemCount;
  final OnTap onTap;
  final Color color1;
  final Color color2;

  const GridWidget({
    Key key,
    this.builder,
    this.itemCount,
    this.onTap,
    this.color1,
    this.color2,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return GridView.count(
      crossAxisCount: 2,
      children: List.generate(itemCount, (index) {
        final line = ((index + 1) / 2).ceil();
        Color color;
        if (line % 2 == 0) {
          color = index % 2 == 0 ? color1 : color2;
        } else {
          color = index % 2 == 0 ? color2 : color1;
        }
        return Material(
          color: color,
          child: InkWell(
            onTap: onTap == null ? null : () => onTap(index),
            child: Center(
              child: builder(
                context,
                index,
                CrazyListLayout.grid,
              ),
            ),
          ),
        );
      }),
    );
  }
}


lib/src/list.dart
part of '../crazy_list.dart';

class ListWidget extends StatefulWidget {
  final WidgetBuilder builder;
  final int itemCount;
  final OnTap onTap;
  final Color color1;
  final Color color2;

  const ListWidget({Key key, this.builder, this.itemCount, this.onTap, this.color1, this.color2}) : super(key: key);

  @override
  _ListWidgetState createState() => _ListWidgetState();
}

class _ListWidgetState extends State<ListWidget> {
  final GlobalKey<AnimatedListState> _key = GlobalKey();
  var _itemCount = 0;

  @override
  void initState() {
    super.initState();
    SchedulerBinding.instance.addPostFrameCallback((_) {
      _animItems();
    });
  }

  @override
  void didUpdateWidget(ListWidget oldWidget) {
    if (oldWidget.itemCount != widget.itemCount) {
      _itemCount = widget.itemCount;
    }
    super.didUpdateWidget(oldWidget);
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedList(
      key: _key,
      itemBuilder: (context, index, animation) => SlideTransition(
            position: Tween<Offset>(
              begin: const Offset(1.0, .0),
              end: Offset.zero,
            )
                .chain(CurveTween(
                  curve: Curves.fastOutSlowIn,
                ))
                .animate(animation),
            child: Card(
              color: index % 2 == 0 ? widget.color1 : widget.color2,
              child: InkWell(
                  onTap: widget.onTap == null ? null : () => widget.onTap(index), child: Center(child: widget.builder(context, index, CrazyListLayout.list))),
            ),
          ),
      initialItemCount: _itemCount,
    );
  }

  void _animItems() async {
    for (var i = 0; i < widget.itemCount; i++) {
      if (i > 0) {
        await Future.delayed(Duration(milliseconds: 100));
      }
      _key.currentState.insertItem(_itemCount);
      _itemCount++;
    }
  }
}


lib/crazy_list.dart
library infinite_widgets;

import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';

part 'src/diagonal.dart';
part 'src/grid.dart';
part 'src/list.dart';
part 'src/single.dart';
part 'src/square.dart';
part 'src/triangle.dart';

typedef WidgetBuilder = Widget Function(BuildContext context, int index, CrazyListLayout mode);
typedef OnTap = Function(int);

enum CrazyListLayout { single, diagonal, triangle, square, list, grid }

enum CrazyListMode { list, grid }

class CrazyList extends StatelessWidget {
  final WidgetBuilder builder;
  final int itemCount;
  final OnTap onTap;
  final Curve curve;
  final Color color1;
  final Color color2;
  final CrazyListMode mode;

  const CrazyList({
    Key key,
    this.builder,
    this.itemCount,
    this.onTap,
    this.mode = CrazyListMode.list,
    this.curve = Curves.fastLinearToSlowEaseIn,
    this.color1,
    this.color2,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    if (itemCount == 1) {
      return SingleWidget(builder: builder);
    } else if (itemCount == 2) {
      return DiagonalWidget(
        builder: builder,
        curve: curve,
        color1: color1,
        color2: color2,
        onTap: onTap,
      );
    } else if (itemCount == 3) {
      return TriangleWidget(
        builder: builder,
        curve: curve,
        color1: color1,
        color2: color2,
        onTap: onTap,
      );
    } else if (itemCount == 4) {
      return SquareWidget(
        builder: builder,
        curve: curve,
        color1: color1,
        color2: color2,
        onTap: onTap,
      );
    }

    if (mode == CrazyListMode.grid) {
      return GridWidget(
        color2: color2,
        color1: color1,
        onTap: onTap,
        builder: builder,
        itemCount: itemCount,
      );
    }
    return ListWidget(
      color2: color2,
      color1: color1,
      onTap: onTap,
      builder: builder,
      itemCount: itemCount,
    );
  }
}

enum _Position { TOP_LEFT, TOP_RIGHT, OUTER_TOP_RIGHT, BOTTOM_LEFT, BOTTOM_RIGHT, OUTER_BOTTOM_RIGHT }

class _TriangleClipper extends CustomClipper<Path> {
  @override
  Path getClip(Size size) {
    var path = Path();
    path.lineTo(0.0, size.height);
    path.lineTo(size.width, size.height / 2);
    path.close();
    return path;
  }

  @override
  bool shouldReclip(CustomClipper<Path> oldClipper) {
    return false;
  }
}

class _DiagonalClipper extends CustomClipper<Path> {
  final double clipHeight;
  final _Position position;

  _DiagonalClipper(this.clipHeight, this.position);

  @override
  Path getClip(Size size) {
    switch (position) {
      case _Position.TOP_LEFT:
        return _getTopLeftPath(size);
      case _Position.OUTER_TOP_RIGHT:
        return _getOuterTopRightPath(size);
      case _Position.TOP_RIGHT:
        return _getTopRightPath(size);
      case _Position.BOTTOM_LEFT:
        return _getBottomLeftPath(size);
      case _Position.BOTTOM_RIGHT:
        return _getBottomRightPath(size);
      case _Position.OUTER_BOTTOM_RIGHT:
        return _getOuterBottomRightPath(size);
      default:
        return _getBottomLeftPath(size);
    }
  }

  @override
  bool shouldReclip(CustomClipper<Path> oldClipper) {
    return false;
  }

  _getTopLeftPath(Size size) {
    var path = Path();
    path.lineTo(0.0, size.height);
    path.lineTo(size.width, size.height);
    path.lineTo(size.width, clipHeight);
    path.close();
    return path;
  }

  _getTopRightPath(Size size) {
    var path = Path();
    path.moveTo(0.0, clipHeight);
    path.lineTo(0.0, size.height);
    path.lineTo(size.width, size.height);
    path.lineTo(size.width, 0.0);
    path.close();
    return path;
  }

  _getOuterTopRightPath(Size size) {
    var path = Path();
    path.lineTo(0.0, 0.0);
    path.lineTo(size.width, 0.0);
    path.lineTo(size.width, clipHeight);
    path.close();
    return path;
  }

  _getOuterBottomRightPath(Size size) {
    var path = Path();
    path.moveTo(0.0, size.height);
    path.lineTo(size.width, size.height);
    path.lineTo(size.width, clipHeight);
    path.close();
    return path;
  }

  _getBottomLeftPath(Size size) {
    var path = Path();
    path.lineTo(0.0, size.height);
    path.lineTo(size.width, size.height - clipHeight);
    path.lineTo(size.width, 0.0);
    path.close();
    return path;
  }

  _getBottomRightPath(Size size) {
    var path = Path();
    path.lineTo(0.0, size.height - clipHeight);
    path.lineTo(size.width, size.height);
    path.lineTo(size.width, 0.0);
    path.close();
    return path;
  }
}

class MaterialInkWell extends StatelessWidget {
  final Widget child;
  final VoidCallback onTap;

  const MaterialInkWell({Key key, this.onTap, this.child}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Material(
      child: InkWell(
        child: child,
        onTap: onTap,
      ),
      color: Colors.transparent,
    );
  }
}


lib/src/square.dart
part of '../crazy_list.dart';

class SquareWidget extends StatefulWidget {
  final WidgetBuilder builder;
  final OnTap onTap;
  final Curve curve;
  final Color color1;
  final Color color2;

  const SquareWidget({Key key, this.builder, this.onTap, this.curve, this.color1, this.color2}) : super(key: key);

  @override
  _SquareWidgetState createState() => _SquareWidgetState();
}

class _SquareWidgetState extends State<SquareWidget> with SingleTickerProviderStateMixin {
  AnimationController _animationController;
  Animation _animation;

  @override
  void initState() {
    _animationController = AnimationController(vsync: this);
    _animationController.duration = Duration(milliseconds: 800);
    _animation = CurvedAnimation(parent: _animationController, curve: widget.curve);
    _animationController.forward();
    super.initState();
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      mainAxisSize: MainAxisSize.max,
      children: <Widget>[
        Expanded(
          child: Row(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            mainAxisSize: MainAxisSize.max,
            children: <Widget>[
              Expanded(
                child: AnimatedBuilder(
                  animation: _animation,
                  builder: (context, child) => FractionalTranslation(
                        translation: Offset(-1 + _animation.value, -1 + _animation.value),
                        child: child,
                      ),
                  child: Container(
                    color: widget.color1,
                    width: double.infinity,
                    height: double.infinity,
                    child: MaterialInkWell(
                      onTap: () => widget.onTap(0),
                      child: Center(child: widget.builder(context, 0, CrazyListLayout.square)),
                    ),
                  ),
                ),
              ),
              Expanded(
                child: AnimatedBuilder(
                  animation: _animation,
                  builder: (context, child) => FractionalTranslation(
                        translation: Offset(1 - _animation.value, -1 + _animation.value),
                        child: child,
                      ),
                  child: Container(
                    color: widget.color2,
                    width: double.infinity,
                    height: double.infinity,
                    child: MaterialInkWell(
                      onTap: () => widget.onTap(1),
                      child: Center(child: widget.builder(context, 1, CrazyListLayout.square)),
                    ),
                  ),
                ),
              ),
            ],
          ),
        ),
        Expanded(
          child: Row(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            mainAxisSize: MainAxisSize.max,
            children: <Widget>[
              Expanded(
                child: AnimatedBuilder(
                  animation: _animation,
                  builder: (context, child) => FractionalTranslation(
                        translation: Offset(-1 + _animation.value, 1 - _animation.value),
                        child: child,
                      ),
                  child: Container(
                    color: widget.color2,
                    width: double.infinity,
                    height: double.infinity,
                    child: MaterialInkWell(
                      onTap: () => widget.onTap(2),
                      child: Center(child: widget.builder(context, 2, CrazyListLayout.square)),
                    ),
                  ),
                ),
              ),
              Expanded(
                child: AnimatedBuilder(
                  animation: _animation,
                  builder: (context, child) => FractionalTranslation(
                        translation: Offset(1 - _animation.value, 1 - _animation.value),
                        child: child,
                      ),
                  child: Container(
                    color: widget.color1,
                    width: double.infinity,
                    height: double.infinity,
                    child: MaterialInkWell(
                      onTap: () => widget.onTap(3),
                      child: Center(child: widget.builder(context, 3, CrazyListLayout.square)),
                    ),
                  ),
                ),
              ),
            ],
          ),
        ),
      ],
    );
  }
}


lib/src/diagonal.dart
part of '../crazy_list.dart';

class DiagonalWidget extends StatefulWidget {
  final WidgetBuilder builder;
  final OnTap onTap;
  final Curve curve;
  final Color color1;
  final Color color2;

  const DiagonalWidget({Key key, this.builder, this.onTap, this.curve, this.color1, this.color2}) : super(key: key);

  @override
  _DiagonalWidgetState createState() => _DiagonalWidgetState();
}

class _DiagonalWidgetState extends State<DiagonalWidget> with SingleTickerProviderStateMixin {
  AnimationController _animationController;
  Animation _animation;

  @override
  void initState() {
    _animationController = AnimationController(vsync: this);
    _animationController.duration = Duration(milliseconds: 800);
    _animation = CurvedAnimation(parent: _animationController, curve: widget.curve);
    _animationController.forward();
    super.initState();
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (context, constraint) {
        final maxHeight = constraint.maxHeight;

        return Stack(
          children: <Widget>[
            AnimatedBuilder(
              animation: _animation,
              builder: (context, child) => FractionalTranslation(
                    translation: Offset((-1 + _animation.value) / 2, (-1 + _animation.value) / 2),
                    child: child,
                  ),
              child: ClipPath(
                clipper: _DiagonalClipper(maxHeight, _Position.BOTTOM_LEFT),
                child: Container(
                  color: widget.color1,
                  width: double.infinity,
                  height: double.infinity,
                  child: MaterialInkWell(
                    onTap: () => widget.onTap(0),
                    child: Align(alignment: Alignment.topLeft, child: widget.builder(context, 0, CrazyListLayout.diagonal)),
                  ),
                ),
              ),
            ),
            AnimatedBuilder(
              animation: _animation,
              builder: (context, child) => FractionalTranslation(
                    translation: Offset((1 - _animation.value) / 2, (1 - _animation.value) / 2),
                    child: child,
                  ),
              child: ClipPath(
                clipper: _DiagonalClipper(maxHeight, _Position.TOP_RIGHT),
                child: Container(
                  color: widget.color2,
                  width: double.infinity,
                  height: double.infinity,
                  child: MaterialInkWell(
                    onTap: () => widget.onTap(1),
                    child: Align(
                      alignment: Alignment.bottomRight,
                      child: widget.builder(context, 1, CrazyListLayout.diagonal),
                    ),
                  ),
                ),
              ),
            )
          ],
        );
      },
    );
  }
}


lib/src/triangle.dart
part of '../crazy_list.dart';

class TriangleWidget extends StatefulWidget {
  final WidgetBuilder builder;
  final OnTap onTap;
  final Curve curve;
  final Color color1;
  final Color color2;

  const TriangleWidget({Key key, this.builder, this.onTap, this.curve, this.color1, this.color2}) : super(key: key);

  @override
  _TriangleWidgetState createState() => _TriangleWidgetState();
}

class _TriangleWidgetState extends State<TriangleWidget> with SingleTickerProviderStateMixin {
  AnimationController _animationController;
  Animation _animation;

  @override
  void initState() {
    _animationController = AnimationController(vsync: this);
    _animationController.duration = Duration(milliseconds: 800);
    _animation = CurvedAnimation(parent: _animationController, curve: widget.curve);
    _animationController.forward();
    super.initState();
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (context, constraint) {
        final maxHeight = constraint.maxHeight;

        return Stack(
          children: <Widget>[
            AnimatedBuilder(
              animation: _animation,
              builder: (context, child) => FractionalTranslation(
                    translation: Offset(1 - _animation.value, 0),
                    child: child,
                  ),
              child: ClipPath(
                clipper: _DiagonalClipper(maxHeight / 2, _Position.OUTER_TOP_RIGHT),
                child: Container(
                  color: widget.color2,
                  width: double.infinity,
                  height: double.infinity,
                  child: MaterialInkWell(
                    onTap: () => widget.onTap(0),
                    child: Align(alignment: Alignment.topRight, child: widget.builder(context, 0, CrazyListLayout.triangle)),
                  ),
                ),
              ),
            ),
            AnimatedBuilder(
              animation: _animation,
              builder: (context, child) => FractionalTranslation(
                    translation: Offset(-1 + _animation.value, 0),
                    child: child,
                  ),
              child: ClipPath(
                clipper: _TriangleClipper(),
                child: Container(
                  color: widget.color1,
                  width: double.infinity,
                  height: double.infinity,
                  child: MaterialInkWell(
                    onTap: () => widget.onTap(1),
                    child: Center(child: widget.builder(context, 1, CrazyListLayout.triangle)),
                  ),
                ),
              ),
            ),
            AnimatedBuilder(
              animation: _animation,
              builder: (context, child) => FractionalTranslation(
                    translation: Offset(1 - _animation.value, 0),
                    child: child,
                  ),
              child: ClipPath(
                clipper: _DiagonalClipper(maxHeight / 2, _Position.OUTER_BOTTOM_RIGHT),
                child: Container(
                  color: widget.color2,
                  width: double.infinity,
                  height: double.infinity,
                  child: MaterialInkWell(
                    onTap: () => widget.onTap(2),
                    child: Align(alignment: Alignment.bottomRight, child: widget.builder(context, 2, CrazyListLayout.triangle)),
                  ),
                ),
              ),
            ),
          ],
        );
      },
    );
  }
}


example/lib/main.dart
import 'dart:core';

import 'package:crazy_list/crazy_list.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  var itemCount = 7;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Column(
        children: <Widget>[
          Expanded(
            child: CrazyList(
              itemCount: itemCount,
              mode: CrazyListMode.list,
              color1: Colors.green[50],
              color2: Colors.green[300],
              onTap: (index) => print('$index tapped'),
              builder: (context, index, mode) {
                return Padding(
                  padding: const EdgeInsets.all(16.0),
                  child: Text('My Label $index'),
                );
              },
            ),
          ),
          Row(
            mainAxisAlignment: MainAxisAlignment.spaceEvenly,
            children: <Widget>[
              FlatButton(
                child: Text('-1'),
                onPressed: () {
                  setState(() {
                    if (itemCount > 1) {
                      itemCount--;
                    }
                  });
                },
              ),
              FlatButton(
                child: Text('+1'),
                onPressed: () {
                  setState(() {
                    itemCount++;
                  });
                },
              ),
            ],
          ),
        ],
      ),
    );
  }
}
